Implement homogeneous heuristic

Erotemic opened the pull request:
@edgarsi If you don't mind taking a look, I'd appreciate a review.


To get the worst-case overhead benchmark I tested the overhead when iterating over nothing. Compared to the baseline, the current release on pypi has a ~5x slowdown and the new homogeneous=True as well as the homogeneous='auto' heuristic behavior matches it. When homogeneous=False, the slowdown is ~100x.

I've hard coded the parameters for the homogeneous auto heuristic. It currently checks the first 5 iterations for any iteration that takes more than 5e-4 seconds, and continues on the slow path, otherwise it breaks into the fast path, at which point a drastic slowdown in progress iteration will cause the failure mode homogeneous=False fixes.

I did abstract some of the boilerplate into a function. I haven't tested if that has a speed impact, I'm assuming one extra function call will be ok.

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+                    yield item
+                    self._slow_path_step_body()
+
+            except _GottaGoFast:
```
Branching using exceptions is not needed.
```
if homogeneous:
    slow_path_loop()
else:
    use_fast_path = True
```

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+
+            except _GottaGoFast:
+                use_fast_path = True
+            except StopIteration:
```
Not sure what this achieves.

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+                    overhead_threshold = 50e-9 * 10_000
+
+                    _check_times = []
+                    for (self._iter_idx, item), _ in zip(gen, range(num_initial_steps)):
```
Using `zip` causes the 5 elements to be generated and stored in memory, breaking the support for cases where there is only enough memory for one element at a time. `itertools.islice` has been in the standard library forever(?), so is probably ok to use.

Erotemic replied:
Good call.

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+
+                    if len(_check_times) > 1:
+                        # smallest, *middle, slowest = sorted(_check_times)
+                        slowest = max(_check_times)
```
max?

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+                # Slow path where we do checks every iteration.
+                for self._iter_idx, item in gen:
+                    yield item
+                    self._slow_path_step_body()
```
I'm certain a function call is slower than checking `self._iter_idx == num_initial_steps` in each loop. But there is no need for either drawback.
```
def slow_path_loop(gen):
    ...

for i in range(num_initial_steps):
    slow_path_loop(islice(gen, 1))
    _check_times.append(self.between_time)
```

edgarsi replied:
The point being that it is fine to add what are effectively constant overheads during the first 5 iterations, but the overheads do matter during the loop where most items are iterated.

edgarsi commented on progiter/progiter.py:
```diff
@@ -470,29 +485,74 @@ def _iterate(self):
+                        self._slow_path_step_body()
+                        _check_times.append(self._between_time)
+
+                    if len(_check_times) > 1:
```
This is more complex than needs to be.
```
slowest = 0
for ... range(num_initial_steps):
    ...
    slowest = max(slowest, self._between_time)

if slowest < overhead_threshold:
    homogenous = True
```

Erotemic commented:
I noticed another bug where sometimes the end message is not displayed. This can happen if the estimates are updated but the display step is skipped.

I've added a new internal tracking variable `_display_idx` to track the most recent iteration a message was displayed on. While I was doing this I refactored the other internal variables `_now_idx`, `_now_time`, `_prev_idx` and `_prev_time`, which were poorly named. These are now called `_measured_idx`, `_measured_time`, and `_prev_measured_idx`, and `_prev_measured_time`, which I think makes more sense.
